src/mlpack/tests/main_tests/kfn_test.cpp

```diff
@@ -0,0 +1,378 @@
+ */
+BOOST_AUTO_TEST_CASE(KFNInvalidEpsilonTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("epsilon", (double) -1); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Make sure that dimensions of the neighbors and distances
+ * matrices are correct given a value of k.  
+ */
+BOOST_AUTO_TEST_CASE(KFNOutputDimensionTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+  
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  
+  mlpackMain();
+
+  // Check the neighbors matrix has 4 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>("neighbors").n_rows, 10);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>("neighbors").n_cols, 100);
+
+  // Check the distances matrix has 4 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_rows, 10);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_cols, 100);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(KFNModelReuseTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  arma::mat queryData;
+  queryData.randu(3, 90); // 90 points in 3 dimensions.
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("query", queryData);
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighbors = std::move(CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  distances = std::move(CLI::GetParam<arma::mat>("distances"));
+  
+  // Reset passed parameters. 
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+
+  // Input saved model, pass the same query and keep k unchanged.
+  SetInputParam("input_model", 
+      std::move(CLI::GetParam<KFNModel*>("output_model")));
+  SetInputParam("query", queryData);
+  
+  mlpackMain();
+
+  // Check that initial output matrices and the output matrices using 
+  // saved model are equal.
+  CheckMatrices(neighbors, CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  CheckMatrices(distances, CLI::GetParam<arma::mat>("distances"));
+}
+
+/*
+ * Ensure that different search algorithms give same result.
+ */
+BOOST_AUTO_TEST_CASE(KFNAllAlgorithmsTest)
+{ 
+  string algorithms[] = {"dual_tree", "naive", "single_tree", "greedy"};
+  int nof_algorithms = sizeof(algorithms)/sizeof(algorithms[0]);
```

Comment:
This is a minor style comment, but can you make sure that variable names don't have underscores in them like the style guide specifies?  https://github.com/mlpack/mlpack/wiki/DesignGuidelines#style-guidelines

So `nofAlgorithms`, for example.  I think there are other places in this code where the same issue applies, so if you can take a look through and fix it I'd appreciate it.  Also horizontal whitespace in this line: `sizeof(algorithms) / sizeof(algorithms[0])`.